garak/generators/openai.py

```diff
@@ -223,6 +224,32 @@ def _call_model(
             if hasattr(self, arg) and arg not in self.suppressed_params:
                 create_args[arg] = getattr(self, arg)
 
+        # basic token boundary validation to ensure requests are not rejected for exceeding target context length
+        generation_max_tokens = create_args.get("max_tokens", None)
+        if generation_max_tokens is not None:
+            # count tokens in prompt and ensure max_tokens requested is <= context_len allowed
+            if (
+                hasattr(self, "context_len")
+                and self.context_len is not None
+                and generation_max_tokens > self.context_len
+            ):
+                logging.warning(
+                    f"Requested max_tokens {generation_max_tokens} exceeds context length {self.context_len}, reducing requested maximum"
+                )
+                generation_max_tokens = self.context_len
+            prompt_tokens = 0
+            try:
+                encoding = tiktoken.encoding_for_model(self.name)
+                prompt_tokens = len(encoding.encode(prompt))
+            except KeyError as e:
+                prompt_tokens = len(prompt.split())  # extra naive fallback
```

Comment:
LLMs tend to use [subword tokenisation](https://aclanthology.org/P16-1162.pdf) which means word count is a lower bound for token count. Space splitting also adds punctuation in to words, and that also deflates token count. Typical practice here is to use a fudge factor multiplier or divider based on whitespace or character count. See https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them

Reply:
Maybe go with the `3/4` of a word per token factor then?
 ```suggestion
                prompt_tokens = len(prompt.split()) * 4 / 3  # extra naive fallback 1 token ~= 3/4 of a word
```

This may also grossly overestimate the prompt token count, though for common models the `tiktoken` value should have been available.

Reply:
Yeah, tiktoken would be more precise, but I don't think it wins in the speed/need for precision tradeoff or in the install weight/need for precision tradeoff.

Prefer char fudge factor to whitespace - the former is more resilient to variation in mean word length